snekbox/snekio.py

```diff
@@ -0,0 +1,108 @@
+"""I/O Operations for sending / receiving files from the sandbox."""
+from __future__ import annotations
+
+import zlib
+from base64 import b64decode, b64encode
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Generic, TypeVar
+
+RequestType = dict[str, str | bool | list[str | dict[str, str]]]
+
+T = TypeVar("T", str, bytes)
+
+
+def sizeof_fmt(num: int, suffix: str = "B") -> str:
+    """Return a human-readable file size."""
+    for unit in ("", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"):
+        if abs(num) < 1024:
+            return f"{num:3.1f}{unit}{suffix}"
+        num /= 1024
+    return f"{num:.1f}Yi{suffix}"
+
+
+class AttachmentError(ValueError):
+    """Raised when an attachment is invalid."""
+
+
+class ParsingError(AttachmentError):
+    """Raised when an incoming file cannot be parsed."""
+
+
+class IllegalPathError(AttachmentError):
+    """Raised when an attachment has an illegal path."""
+
+
+@dataclass
+class FileAttachment(Generic[T]):
+    """A file attachment."""
+
+    name: str
+    content: T
+
+    @classmethod
+    def from_dict(cls, data: dict[str, str]) -> FileAttachment:
+        """Convert a dict to an attachment."""
+        name = data["name"]
+        path = Path(name)
+        parts = path.parts
+
+        if path.is_absolute() or set(parts[0]) & {"\\", "/"}:
+            raise IllegalPathError(f"File path '{name}' must be relative")
+
+        if any(set(part) == {"."} for part in parts):
+            raise IllegalPathError(f"File path '{name}' may not use traversal ('..')")
```

Comment:
This only checks a single dot. Shouldn't it also check `..`? I wonder if there is a more convenient way to detect this. Maybe https://stackoverflow.com/a/45190125 ?

Reply:
Implemented the pathlib solution since [00b7e3b](https://github.com/python-discord/snekbox/pull/159/commits/00b7e3bd13b7914a44c5af00ff6f8da9b48fa9c5) as detailed in the stackoverflow post. Should be better since now we only disallow traversal beyond root.